File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/HistogramFixedWidth.java
/*******************************************************************************
 * Copyright (c) 2015-2018 Skymind, Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

package org.nd4j.linalg.api.ops.impl.transforms;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.common.base.Preconditions;
import org.nd4j.imports.NoOpNameFoundException;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ops.DynamicCustomOp;
import org.tensorflow.framework.AttrValue;
import org.tensorflow.framework.GraphDef;
import org.tensorflow.framework.NodeDef;

import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * Histogram fixed with op
 *
 * @author Alex Black
 */
public class HistogramFixedWidth extends DynamicCustomOp {

    public HistogramFixedWidth(SameDiff sameDiff, SDVariable values, SDVariable valuesRange, SDVariable numBins) {
        super(sameDiff, numBins == null ? new SDVariable[]{values, valuesRange} : new SDVariable[]{values, valuesRange, numBins}, false);
    }

    public HistogramFixedWidth() {
        //no-op
    }

    @Override
    public String opName() {
        return "histogram_fixed_width";
    }


    @Override
    public String onnxName() {
        throw new NoOpNameFoundException("No onnx op opName found for " +  opName());
    }

    @Override
    public String tensorflowName() {
        return "HistogramFixedWidth";
    }

    @Override
    public void initFromTensorFlow(NodeDef nodeDef, SameDiff initWith, Map<String, AttrValue> attributesForNode, GraphDef graph) {
        //No op - just need the inputs
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> f1) {
        throw new UnsupportedOperationException("Not supported");
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes){
        //1 or 2 possible: 2 for TF import (fill with specified value
        Preconditions.checkState(dataTypes != null && (dataTypes.size() == 2 || dataTypes.size() == 3),
                "Expected 2 or 3 input datatypes for %s, got %s", getClass(), dataTypes);
        //TODO MAKE CONFIGURABLE
        return Collections.singletonList(DataType.INT);
    }
}

File: libnd4j/include/ops/declarable/helpers/cpu/histogramFixedWidth.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// @author Yurii Shyrma (iuriish@yahoo.com), created on 31.08.2018
//
#include <ops/declarable/helpers/histogramFixedWidth.h>
#if NOT_EXCLUDED(OP_histogram_fixed_width)
namespace sd {
namespace ops {
namespace helpers {

template <typename T>
void histogramFixedWidth_(NDArray& input, NDArray& range, NDArray& output) {
  const int nbins = output.lengthOf();

  // firstly initialize output with zeros
  output.nullify();

  const T leftEdge = static_cast<T>(range.e<double>(0));
  const T rightEdge = static_cast<T>(range.e<double>(1));

  const T binWidth = (rightEdge - leftEdge) / nbins;
  const T secondEdge = leftEdge + binWidth;
  const T lastButOneEdge = rightEdge - binWidth;

  sd::LongType inputLength = input.lengthOf();

  // FIXME: make this one parallel without CRITICAL section
  for (sd::LongType i = 0; i < inputLength; ++i) {
    const T value = input.e<T>(i);

    if (value < secondEdge) {
      output.p<sd::LongType>(0, output.e<sd::LongType>(0) + 1);
    } else if (value >= lastButOneEdge) {
      output.p<sd::LongType>(nbins - 1, output.e<sd::LongType>(nbins - 1) + 1);
    } else {
      sd::LongType currInd = static_cast<sd::LongType>((value - leftEdge) / binWidth);
      output.p<sd::LongType>(currInd, output.e<sd::LongType>(currInd) + 1);
    }
  }
}

void histogramFixedWidth(sd::LaunchContext* context, NDArray& input, NDArray& range, NDArray& output) {
  BUILD_SINGLE_SELECTOR(input.dataType(), histogramFixedWidth_, (input, range, output), SD_COMMON_TYPES);
}
BUILD_SINGLE_TEMPLATE( void histogramFixedWidth_, (NDArray& input, NDArray& range, NDArray& output),
                      SD_COMMON_TYPES);

}  // namespace helpers
}  // namespace ops
}  // namespace sd
#endif